jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/sys/FusekiAutoModules.java

```diff
@@ -0,0 +1,230 @@
+    }
+
+    // Single auto module controller.
+    private static FusekiServiceLoaderModules autoModules = null;
+
+    private static FusekiServiceLoaderModules getServiceLoaderModules() {
+        if ( autoModules == null )
+            setup();
+        return autoModules;
+    }
+
+    private static FusekiServiceLoaderModules createServiceLoaderModules() {
+        FusekiServiceLoaderModules newAutoModules = new FusekiServiceLoaderModules();
+        newAutoModules.setDiscovery();
+        return newAutoModules;
+    }
+
+    /**
+     * Use {@link java.util.ServiceLoader} to find {@link FusekiModule}
+     * available via the classpath or modules.
+     * <p>
+     * These are the modules used when building a {@link FusekiServer} if
+     * {@link FusekiServer.Builder#setFusekiModules} is not used.
+     */
+    private static class FusekiServiceLoaderModules {
+
+        // This keeps the list of discovered Fuseki modules.
+        private ServiceLoader<FusekiAutoModule> serviceLoader = null;
+
+        private FusekiServiceLoaderModules() { }
+
+        private void setDiscovery() {
+            serviceLoader = discover();
+        }
+
+        /**
+         * Discover FusekiModules via {@link java.util.ServiceLoader}.
+         * This step does not create the module objects.
+         */
+        private ServiceLoader<FusekiAutoModule> discover() {
+            // Look for the 4.8.0 name (FusekiModule) which (4.9.0) is split into
+            // FusekiModule (interface) and FusekiAutoModule (this is loaded by ServiceLoader)
+            // Remove sometime!
+            discoveryWarnLegacy();
+
+            Class<FusekiAutoModule> moduleClass = FusekiAutoModule.class;
+            ServiceLoader<FusekiAutoModule> newServiceLoader = null;
+            synchronized (this) {
+                try {
+                    newServiceLoader = ServiceLoader.load(moduleClass, this.getClass().getClassLoader());
+                } catch (ServiceConfigurationError ex) {
+                    FmtLog.error(LOG, ex, "Problem with service loading for %s", moduleClass.getName());
+                    throw ex;
+                }
+                if ( LOG.isDebugEnabled() ) {
+                    newServiceLoader.stream().forEach(provider->{
+                        FmtLog.info(LOG, "Fuseki Module: %s", provider.type().getSimpleName());
+                    });
+                }
+            }
+            return newServiceLoader;
+        }
+
+        private void discoveryWarnLegacy() {
+            Class<FusekiModule> moduleClass = FusekiModule.class;
+            try {
+                ServiceLoader<FusekiModule> newServiceLoader = ServiceLoader.load(moduleClass, this.getClass().getClassLoader());
+                newServiceLoader.stream().forEach(provider->{
+                    FmtLog.warn(FusekiAutoModules.class, "Legacy: \"%s\" : Interface FusekiModule changed to FusekiAutoModule", provider.type().getSimpleName());
```

Comment:
Might want to be explicit in this warning that the legacy module is ignored as a result

Reply:
Done